proposals/0106-paritioned-epoch-rewards-pda.md

```diff
@@ -0,0 +1,100 @@
+authors:
+  - Tyera Eulberg
+category: Standard
+type: Core
+status: Draft
+created: 2024-01-17
+feature: https://github.com/solana-labs/solana/issues/32166
+extends: 0015
+---
+
+## Summary
+
+Partitioned epoch rewards (SIMD 0015) will split epoch rewards distribution
+across multiple blocks at the start of an epoch. This extension of that SIMD
+proposes storing data about the rewards calculation and partitions in a sysvar
+account at a program-derived address (PDA), which can be queried by clients, or
+indeed within the runtime, to performantly identify the partition for the
+distribution to any particular address.
+
+## Motivation
+
+When we move to partitioned epoch rewards as described in the original SIMD, the
+only way to find the stake or voting rewards for a specific address will be to
+iterate through blocks at the beginning of each epoch, inspecting all the
+rewards. This is because the runtime does not persist information about how the
+rewards were partitioned; in fact, it does not even persist how many blocks the
+rewards distribution spans, so there is no way to predict how long it will take
+or how far through the epoch to go in finding a particular address.
+
+## Alternatives Considered
+
+An alternative to storing partition data in an on-chain account would be to
+record the necessary data in the ledger in some fashion. This could be a
+transaction that gets added to the block, a special RewardType, or metadata that
+gets stored on the node and then duplicated to long-term storage, like Bigtable.
+In fact, it is probably worthwhile to pursue this alternative as well, since it
+will enable finding rewards without access to snapshots or the running chain.
+
+## New Terminology
+
+None
+
+## Detailed Design
+
+When partitioned rewards are calculated in the runtime (currently during the
+first block of the epoch), the runtime should populate a PDA that stores the
+partition data needed to recreate the hasher that returns the partition index
+for any address. This data comprises: the number of partitions, the parent
+blockhash, and the hasher kind. More specifically:
+
+```rust
+// Version wrapper to allow future updates
+// Variant serialized as little-endian unsigned 32-bit integer
+enum EpochRewardsPartitionDataVersion {
+    V0(PartitionData),
+}
+
+// Extensible list of kinds of hashers used to generate partitions
+// Mapping fn defines the HasherKind for each partition-data version
+enum HasherKind {
+    Sip13 = 0,
+}
+
+// Data about rewards partitions for a particular epoch
+struct PartitionData {
+    num_partitions: usize, // little-endian unsigned 64-bit integer
+    parent_blockhash: Hash, // byte-array of length 32
+}
+```
+
+The address of this PDA will use some bytes -- to prevent griefing and namespace
+the PDAs -- and the rewards distribution epoch number as a little-endian u64 as
+seeds. Specifically: `[b"EpochRewards",b"PartitionData", &epoch.to_le_bytes()]`.
+The owning program should be the sysvar program id:
```

Comment:
Can you add the reasoning for why it's owned by the sysvar program rather than the stake program? And also comment on how these sysvars will be different from other sysvars? (Writable, for example)

Reply:
I don't actually know what you mean by "(Writable, for example)"

Reply:
I did take a crack at this, but both Trent and I thought it made the proposal worse, so I've reverted it. Here's what I had written:
"Ownership by the Stake program was also considered, but would require either
extending the Stake state enum definitions or for parsers to do special handling
based on data length. Meanwhile, this PDA has more in common with existing
Sysvar-owned accounts: like other sysvars, it is written and controlled only by
the runtime. The only difference from a traditional sysvar is the
program-derived instead of a static address."

Reply:
Maybe worth noting that these sysvars won't be accessible via syscalls since other sysvars typically are assumed to be.